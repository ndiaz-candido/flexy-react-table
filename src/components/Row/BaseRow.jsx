import React, { Component } from 'react';
import PropTypes from 'prop-types';
import {
  LeftSection,
  RightSection,
  ScrollableSection,
} from './sections';

const defaultRowStyle = {
  display: 'flex',
  overflow: 'hidden',
  minHeight: '50px',
};
const stickyStyle = {
  position: 'sticky',
  top: '0',
  zIndex: '1000',
};
const cellDefaultStyles = width => ({
  width: `${width}px`,
  flex: `0 0 ${width}px`,
});

class BaseRow extends Component {
  static DEFAULT_COLUMN_WIDTH = 50;
  static HEADER_ROW_ID = 0;
  static NOT_SPECIFY_COLUMN_VALUE = 'Not Column';

  state = {
    leftStickySection: [],
    rightStickySection: [],
    scrollableSection: [],
  };

  static getDerivedStateFromProps({ columns, rowData, columnClassName }) {
    const leftStickySection = [];
    const rightStickySection = [];
    const scrollableSection = [];

    columns.forEach(column => {
      const columnWidth = column.width || BaseRow.DEFAULT_COLUMN_WIDTH;
      const columnDataKey = column.dataKey || '';
      const cellContent = rowData[columnDataKey] || '';
      const styles = cellDefaultStyles(columnWidth);
      const cell = (
        <div className={columnClassName} style={styles}>
          {cellContent}
        </div>
      );

      switch (column.stickyAlign) {
        case 'left':
          leftStickySection.push(cell);
          break;
        case 'right':
          rightStickySection.push(cell);
          break;
        default:
          scrollableSection.push(cell);
          break;
      }
    });

    return {
      leftStickySection,
      rightStickySection,
      scrollableSection,
    };
  }

  render() {
    const { className, isSticky, onScroll, rowId, scrollLeft } = this.props;
    const {
      leftStickySection,
      rightStickySection,
      scrollableSection,
    } = this.state;
    const stickyStyleProps = isSticky ? stickyStyle : {};
    const styles = {
      ...defaultRowStyle,
      ...stickyStyleProps,
    };

    return (
      <div className={className} style={styles}>
        <LeftSection cells={leftStickySection} />
        <ScrollableSection
          rowId={rowId}
          cells={scrollableSection}
          onScroll={onScroll}
          scrollLeft={scrollLeft}
        />
        <RightSection cells={rightStickySection} />
      </div>
    );
  }
}

BaseRow.propTypes = {
  columns: PropTypes.array,
  className: PropTypes.any,
  columnClassName: PropTypes.any,
  isSticky: PropTypes.bool,
  isHeader: PropTypes.bool,
  onScroll: PropTypes.func,
  rowId: PropTypes.number,
  rowData: PropTypes.object,
  scrollLeft: PropTypes.number,
};

BaseRow.defaultProps = {
  columns: [],
  className: '',
  columnClassName: '',
  isSticky: false,
  isHeader: false,
  onScroll: () => {},
  rowId: -1,
  rowData: {},
  scrollLeft: 0,
};

export default BaseRow;
