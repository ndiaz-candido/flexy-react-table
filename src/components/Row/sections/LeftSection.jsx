import React, { Fragment } from 'react';
import PropTypes from 'prop-types';

/* eslint-disable react/no-array-index-key */

const defaultSectionStyle = {
  display: 'flex',
  flex: '0 0 auto',
  position: 'sticky',
  left: '0',
};
const LeftSection = ({ cells }) => (
  <div style={defaultSectionStyle}>
    {cells.map((cell, index) => (
      <Fragment key={`leftSection-${index}`}>{cell}</Fragment>
    ))}
  </div>
);

LeftSection.propTypes = {
  cells: PropTypes.array,
};

LeftSection.defaultProps = {
  cells: [],
};

export default LeftSection;
