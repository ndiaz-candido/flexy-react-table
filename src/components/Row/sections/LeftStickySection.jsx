import React, { Fragment } from 'react';
import PropTypes from 'prop-types';

const defaultSectionStyle = {
  display: 'flex',
  flex: '0 0 auto',
  position: 'sticky',
  left: '0',
};
const LeftStickySection = ({ cells }) => (
  <div style={defaultSectionStyle}>
    { cells.map((cell, index) => <Fragment key={`leftSection-${index}`}>{cell}</Fragment>) }
  </div>
);

LeftStickySection.propTypes = {
  cells: PropTypes.array,
};

LeftStickySection.defaultProps = {
  cells: []
};

export default LeftStickySection;
