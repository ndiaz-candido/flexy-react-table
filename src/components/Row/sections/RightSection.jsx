import React, { Fragment } from 'react';
import PropTypes from 'prop-types';

/* eslint-disable react/no-array-index-key */

const defaultSectionStyle = {
  display: 'flex',
  flex: '0 0 auto',
  position: 'sticky',
  right: '0',
};
const RightStickySection = ({ cells }) => (
  <div style={defaultSectionStyle}>
    {cells.map((cell, index) => (
      <Fragment key={`rightSection=${index}`}>{cell}</Fragment>
    ))}
  </div>
);

RightStickySection.propTypes = {
  cells: PropTypes.array.isRequired,
};

export default RightStickySection;
