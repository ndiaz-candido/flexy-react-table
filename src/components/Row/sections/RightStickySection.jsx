import React, { Fragment } from 'react';
import PropTypes from 'prop-types';

const defaultSectionStyle = {
  display: 'flex',
  flex: '0 0 auto',
  position: 'sticky',
  right: '0',
};
const RightStickySection = ({ cells }) => (
  <div style={defaultSectionStyle}>
    {cells.map((cell, index) => (<Fragment key={`rightSection-${index}`}>{cell}</Fragment>))}
  </div>
);

RightStickySection.propTypes = {
  cells: PropTypes.array,
};

RightStickySection.defaultProps = {
  cells: []
};

export default RightStickySection;
