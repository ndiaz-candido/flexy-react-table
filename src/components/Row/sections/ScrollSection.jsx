import React, { PureComponent, Fragment } from 'react';
import PropTypes from 'prop-types';
import { Scrollbars } from 'react-custom-scrollbars';

/* eslint-disable react/no-array-index-key */
/* eslint-disable react/no-multi-comp */

const defaultWrapperStyle = {
  flex: '1 1 auto',
  position: 'relative',
};
const defaultSectionStyle = {
  display: 'flex',
};
const defaultScrollRightIndicatorStyle = {
  height: '100%',
  position: 'absolute',
  right: 0,
  width: '20px',
  backgroundImage:
    'linear-gradient(to right, rgba(202, 203, 204, 0), rgba(202, 203, 204, 1))',
};
const defaultScrollLeftIndicatorStyle = {
  height: '100%',
  position: 'absolute',
  left: 0,
  width: '20px',
  backgroundImage:
    'linear-gradient(to left, rgba(202, 203, 204, 0), rgba(202, 203, 204, 1))',
};

class ScrollSection extends PureComponent {
  state = {
    showScrollTrack: false,
  };

  componentDidMount() {
    const { rowId, registerScrollSectionElements, onScroll } = this.props;

    registerScrollSectionElements(rowId, {
      scrollableArea: this.scrollableAreaRef,
      scrollRightIndicator: this.scrollRightIndicatorRef,
      scrollLeftIndicator: this.scrollLeftIndicatorRef,
    });

    // Call this once on mount to initialize the Table state.
    if (this.scrollableAreaRef) {
      onScroll(this.scrollableAreaRef.getValues());
    }
  }

  handleOnMouseEnter = () => this.setState({ showScrollTrack: true });

  handleOnMouseLeave = () => this.setState({ showScrollTrack: false });

  handleOnScroll = () =>
    this.props.onScroll(this.scrollableAreaRef.getValues());

  renderScrollView = ({ style, ...props }) => {
    const { showScrollTrack } = this.state;
    const showXScrollTrack = { overflowX: 'scroll' };
    const hideXScrollTrack = { overflowX: 'hidden' };
    const customStyles = showScrollTrack ? showXScrollTrack : hideXScrollTrack;

    return (
      <div
        style={{
          ...style,
          ...defaultSectionStyle,
          ...customStyles,
        }}
        {...props}
      />
    );
  };

  setScrollableAreaRef = element => (this.scrollableAreaRef = element);

  setScrollRightIndicatorRef = element =>
    (this.scrollRightIndicatorRef = element);

  setScrollLeftIndicatorRef = element =>
    (this.scrollLeftIndicatorRef = element);

  render() {
    const { cells, showScrollIndicators } = this.props;
    const scrollRightIndicatorStyle = showScrollIndicators
      ? defaultScrollRightIndicatorStyle
      : {};
    const scrollLeftIndicatorStyle = showScrollIndicators
      ? defaultScrollLeftIndicatorStyle
      : {};

    return (
      <div
        id="scrollable-wrapper"
        style={defaultWrapperStyle}
        onMouseEnter={this.handleOnMouseEnter}
        onMouseLeave={this.handleOnMouseLeave}>
        <div
          ref={this.setScrollRightIndicatorRef}
          style={scrollRightIndicatorStyle}
        />
        <div
          ref={this.setScrollLeftIndicatorRef}
          style={scrollLeftIndicatorStyle}
        />
        <Scrollbars
          ref={this.setScrollableAreaRef}
          renderView={this.renderScrollView}
          onScroll={this.handleOnScroll}>
          {cells.map((cell, index) => (
            <Fragment key={`scrollCell-${index}`}>{cell}</Fragment>
          ))}
        </Scrollbars>
      </div>
    );
  }
}

ScrollSection.propTypes = {
  cells: PropTypes.array,
  onScroll: PropTypes.func,
  registerScrollSectionElements: PropTypes.func,
  showScrollIndicators: PropTypes.bool,
};

ScrollSection.defaultProps = {
  cells: [],
  onScroll: () => {},
  registerScrollSectionElements: () => {},
  showScrollIndicators: true,
};

export default ScrollSection;
