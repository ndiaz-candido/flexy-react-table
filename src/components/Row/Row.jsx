import React, { Component } from 'react';
import PropTypes from 'prop-types';
import idGenerator from 'react-id-generator';
import { ScrollSection, RightSection, LeftSection } from './sections';
import ScrollContext from '../ScrollContext';

const defaultRowStyle = {
  display: 'flex',
  minHeight: '50px',
};
const defaultCellStyle = width => ({
  width,
  flex: `0 0 ${width}`,
});

class Row extends Component {
  static DEFAULT_COLUMN_WIDTH = 50;

  state = {
    stickRightSection: [],
    stickLeftSection: [],
    scrollSection: [],
  };

  constructor(props) {
    super(props);

    this.rowId = idGenerator();
  }

  static getDerivedStateFromProps({ columns, cellClassName, rowData }) {
    const stickLeftSection = [];
    const stickRightSection = [];
    const scrollSection = [];

    columns.forEach(column => {
      const columnDataKey = column.dataKey;
      const columnWidth = column.width || Row.DEFAULT_COLUMN_WIDTH;
      const cellContent = rowData[columnDataKey];
      const cell = (
        <div className={cellClassName} style={defaultCellStyle(columnWidth)}>
          {cellContent}
        </div>
      );

      switch (column.stickAlign) {
        case 'right':
          stickRightSection.push(cell);
          break;
        case 'left':
          stickLeftSection.push(cell);
          break;
        default:
          scrollSection.push(cell);
          break;
      }
    });

    return {
      stickLeftSection,
      stickRightSection,
      scrollSection,
    };
  }

  render() {
    const { stickLeftSection, stickRightSection, scrollSection } = this.state;
    const { className } = this.props;

    return (
      <div className={className} style={defaultRowStyle}>
        <LeftSection cells={stickLeftSection} />
        <ScrollContext.Consumer>
          {({ handleOnScroll, registerScrollSection }) => (
            <ScrollSection
              cells={scrollSection}
              onScroll={handleOnScroll}
              registerScrollSection={registerScrollSection}
              rowId={this.rowId}
            />
          )}
        </ScrollContext.Consumer>
        <RightSection cells={stickRightSection} />
      </div>
    );
  }
}

Row.propTypes = {
  className: PropTypes.any,
  cellClassName: PropTypes.any,
  columns: PropTypes.arrayOf(
    PropTypes.shape({
      dataKey: PropTypes.string,
      stickAlign: PropTypes.string,
      width: PropTypes.string,
    })
  ),
  rowData: PropTypes.object,
};

Row.defaultProps = {
  className: '',
  cellClassName: '',
  columns: [],
  rowData: {},
};

export default Row;
