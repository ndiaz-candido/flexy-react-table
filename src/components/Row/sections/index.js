import LeftSection from './LeftSection';
import RightSection from './RightSection';
import ScrollableSection from './ScrollableSection';

export { LeftSection, RightSection, ScrollableSection };
