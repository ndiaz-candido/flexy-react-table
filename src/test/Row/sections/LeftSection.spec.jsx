import React from 'react';
import { shallow } from 'enzyme';
import { LeftSection } from '../../../components/Row/sections';

describe('LeftSection', () => {
  test('should render its default content', () => {
    const LeftSectionComponent = shallow(<LeftSection cells={[]} />);

    expect(LeftSectionComponent).toMatchSnapshot();
  });

  test('should render the given cells', () => {
    const cells = [
      <div key="0">Cell A</div>,
      <div key="1">Cell B</div>,
      <div key="2">Cell C</div>,
    ];
    const LeftSectionComponent = shallow(<LeftSection cells={cells} />);

    expect(LeftSectionComponent).toMatchSnapshot();
  });
});
