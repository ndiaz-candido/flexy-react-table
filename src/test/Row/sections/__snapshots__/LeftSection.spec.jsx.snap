// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LeftSection should render its default content 1`] = `
<div
  style={
    Object {
      "display": "flex",
      "flex": "0 0 auto",
      "left": "0",
      "position": "sticky",
    }
  }
/>
`;

exports[`LeftSection should render the given cells 1`] = `
<div
  style={
    Object {
      "display": "flex",
      "flex": "0 0 auto",
      "left": "0",
      "position": "sticky",
    }
  }
>
  <React.Fragment
    key="leftSection-0"
  >
    <div
      key="0"
    >
      Cell A
    </div>
  </React.Fragment>
  <React.Fragment
    key="leftSection-1"
  >
    <div
      key="1"
    >
      Cell B
    </div>
  </React.Fragment>
  <React.Fragment
    key="leftSection-2"
  >
    <div
      key="2"
    >
      Cell C
    </div>
  </React.Fragment>
</div>
`;
